### PR TITLE
GH-34630: [C++] Second block of refactoring to move acero out of libarrow

### LINKED ISSUE #34630: [C++] Second block of refactoring to move acero out of libarrow
### Describe the enhancement requested

This is the second and last block of refactoring before moving acero out of libarrow.

### Component(s)

C++

### PR DESCRIPTION
**Pull request by ildipo**
### Rationale for this change

This is the second and last block of refactoring before moving acero out of libarrow. This PR removes the remaining dependencies from libarrow into the compute/exec directory. 

### What changes are included in this PR?

In detail:

* compute/exec/expression is moved inside compute since it is needed in several kernels
* all compute/exec specific include files are removed from compute/api.h
* light_array_exec_test.cc i(introduced in the previous refactoring) s merged back into light_array_test
* testing generators are refactored to remove the need for exec/options.h
* some utilities are moved from compute/exec/util to compute/util
* updated python and r libraries usage of moved files

### Are these changes tested?

Yes using existing tests (some of which have been updated due to refactoring).

### Are there any user-facing changes?

no
* Issue: #15280
* Closes: #34630

### CODE REVIEW

File: cpp/src/arrow/compute/exec/plan_test.cc
Comment (icexelloss) [inline cpp/src/arrow/compute/exec/plan_test.cc:1023]: Curious why this change?

**Comment by icexelloss** [comment]
@ildipo High level looks good to me. Is the general idea of moving "expression" from "compute/exec" to "compute" such that it is consider part of "libarrow" and not "libarrow_acero"?

**Comment by icexelloss** [comment]
I also removed most of the reviewers - I think the changes to python/R binding is trivial but if @westonpace think this need review from R maintainers we can add back.

File: cpp/src/arrow/compute/exec/plan_test.cc
Comment (ildipo) [reply cpp/src/arrow/compute/exec/plan_test.cc:1023]: This was using generator.h with compute/exec/options.h. I removed that include (and the associated exec options from generator, that is `SourceNode()`) and expanded it in this test, which was the only place where it was used.

**Comment by ildipo** [comment]
Yes I feel what is in expression.h is used in many other places (e.g. inside kernels) so it belongs to libarrow

**Comment by ildipo** [comment]
I'm not sure why one check failed (apparently it was cancelled???)

**Comment by icexelloss** [comment]
@ildipo Error is this
```
FAILED: src/arrow/compute/CMakeFiles/arrow-compute-internals-test.dir/Unity/unity_0_cxx.cxx.obj 
"C:\Program Files\Git\usr\bin\ccache.exe" C:\PROGRA~2\MICROS~2\2019\ENTERP~1\VC\Tools\MSVC\1429~1.301\bin\Hostx64\x64\cl.exe  /nologo /TP -DARROW_HAVE_RUNTIME_AVX2 -DARROW_HAVE_RUNTIME_AVX512 -DARROW_HAVE_RUNTIME_BMI2 -DARROW_HAVE_RUNTIME_SSE4_2 -DARROW_HAVE_SSE4_2 -DARROW_HDFS -DARROW_MIMALLOC -DARROW_WITH_BENCHMARKS_REFERENCE -DARROW_WITH_LZ4 -DARROW_WITH_RE2 -DARROW_WITH_SNAPPY -DARROW_WITH_UTF8PROC -DARROW_WITH_ZLIB -DARROW_WITH_ZSTD -DBOOST_ALL_NO_LIB -DGMOCK_LINKED_AS_SHARED_LIBRARY=1 -DGTEST_LINKED_AS_SHARED_LIBRARY=1 -DURI_STATIC_BUILD -DUTF8PROC_STATIC -D_CRT_SECURE_NO_WARNINGS -D_ENABLE_EXTENDED_ALIGNED_STORAGE -ID:\a\arrow\arrow\build\cpp\src -ID:\a\arrow\arrow\cpp\src -ID:\a\arrow\arrow\cpp\src\generated -external:ID:\a\arrow\arrow\cpp\thirdparty\flatbuffers\include -external:ID:\a\arrow\arrow\build\cpp\rapidjson_ep\src\rapidjson_ep-install\include -external:ID:\a\arrow\arrow\cpp\thirdparty\hadoop\include -external:ID:\a\arrow\arrow\build\cpp\boost_ep-prefix\src\boost_ep -external:I"C:\Program Files\OpenSSL-Win64\include" -external:ID:\a\arrow\arrow\build\cpp\orc_ep-install\include -external:ID:\a\arrow\arrow\build\cpp\lz4_ep-install\include -external:ID:\a\arrow\arrow\build\cpp\zlib_ep\src\zlib_ep-install\include -external:ID:\a\arrow\arrow\build\cpp\zstd_ep-install\include -external:ID:\a\arrow\arrow\build\cpp\snappy_ep\src\snappy_ep-install\include -external:ID:\a\arrow\arrow\build\cpp\protobuf_ep-install\include -external:ID:\a\arrow\arrow\build\cpp\utf8proc_ep-install\include -external:ID:\a\arrow\arrow\build\cpp\re2_ep-install\include -external:ID:\a\arrow\arrow\build\cpp\xsimd_ep\src\xsimd_ep-install\include -external:ID:\a\arrow\arrow\build\cpp\mimalloc_ep\src\mimalloc_ep\include\mimalloc-2.0 -external:ID:\a\arrow\arrow\build\cpp\googletest_ep-prefix\include -external:W0 /D_SILENCE_TR1_NAMESPACE_DEPRECATION_WARNING /W3 /EHsc /wd5105 /bigobj /utf-8 /W3 /wd4365 /wd4267 /wd4838 /wd4800 /wd4996 /wd4065   /MDd /Z7 /Ob0 /Od /RTC1 /WX -std:c++17 /showIncludes /Fosrc\arrow\compute\CMakeFiles\arrow-compute-internals-test.dir\Unity\unity_0_cxx.cxx.obj /Fdsrc\arrow\compute\CMakeFiles\arrow-compute-internals-test.dir\ /FS -c D:\a\arrow\arrow\build\cpp\src\arrow\compute\CMakeFiles\arrow-compute-internals-test.dir\Unity\unity_0_cxx.cxx
D:\a\arrow\arrow\cpp\src\arrow/compute/exec/test_util.h(49): error C2375: 'arrow::compute::ExecBatchFromJSON': redefinition; different linkage
D:/a/arrow/arrow/cpp/src/arrow/compute/light_array_test.cc(37): note: see declaration of 'arrow::compute::ExecBatchFromJSON'
```

**Comment by icexelloss** [comment]
> Yes I feel what is in expression.h is used in many other places (e.g. inside kernels) so it belongs to libarrow

Got it - makes sense to me. @westonpace does that sound about right?

**Comment by westonpace** [comment]
> Got it - makes sense to me. @westonpace does that sound about right?

Yes, `expression` is pretty stable.  So if the goal is a stable "compute" module then I think moving it inside `compute` should be safe.

Thinking beyond just "stability" I think there is a pretty reasonable separation in Arrow between "compute functions" and "the execution engine" and I think it makes sense for `Expression` to live in the former.  There is also some desire for a Substrait API for expression evaluation.  In theory that could be done without "exec" but that would require splitting the substrait module into two pieces and I don't think that's worth our time (e.g. the expression evaluation API can just depend on "exec" even though it doesn't strictly need to)

**Comment by westonpace** [comment]
I'll try and get a closer look at this tonight but it might be tomorrow.

File: cpp/src/arrow/compute/api.h
Comment (westonpace) [inline cpp/src/arrow/compute/api.h:53]: This could probably be split someday (no need to do it in this PR).  `ExecContext` / `ExecSpan` / `ExecResult` / `ExecValue` / `CallFunction` / `GetFunctionExecutor` belong in compute.  `ExecBatch` belongs in `exec`.  `SelectionVector` could probably be removed at this point and added back in later if we ever decide to do something with it.

File: cpp/src/arrow/compute/exec/plan_test.cc
Comment (westonpace) [reply cpp/src/arrow/compute/exec/plan_test.cc:1023]: I am planning on starting to use these utilities in more places.  However, we can introduce `exec::gen::Gen` (or something like that) which wraps the generator and adds exec-specific utilities.  So I think this is tolerable for now.

File: cpp/src/arrow/compute/type_fwd.h
Comment (westonpace) [inline cpp/src/arrow/compute/type_fwd.h:51]: I think `Declaration` is part of `exec` too.

File: cpp/src/arrow/testing/generator.cc
Comment (westonpace) [inline cpp/src/arrow/testing/generator.cc:373]: Is this needed anywhere?

**Review by westonpace — APPROVED** [review]
A few minor things but this looks correct to me.  Thanks!

File: cpp/src/arrow/compute/api.h
Comment (ildipo) [reply cpp/src/arrow/compute/api.h:53]: I'll fix the other things in this PR and then make another one to move ExecBatch

File: cpp/src/arrow/compute/type_fwd.h
Comment (ildipo) [reply cpp/src/arrow/compute/type_fwd.h:51]: good point, moving in `exec/type_fwd.h`

File: cpp/src/arrow/testing/generator.cc
Comment (ildipo) [reply cpp/src/arrow/testing/generator.cc]: removing

**Comment by ildipo** [comment]
Updated the PR with the changes detailed above.

**Comment by assignUser** [comment]
@github-actions crossbow submit -g cpp

**Comment by icexelloss** [comment]
@westonpace 

There seems to be a failure in recent CI runs (test-alpine-linux-cpp):
```
 68/103 Test  #60: arrow-dataset-file-json-test ..............***Failed   61.16 sec
Running arrow-dataset-file-json-test, redirecting output into /build/cpp/build/test-logs/arrow-dataset-file-json-test.txt (attempt 1/1)
/arrow/cpp/build-support/run-test.sh: line 88: 28113 Segmentation fault      (core dumped) 
```

Also some valgrind failures on the following tests:
https://dev.azure.com/ursacomputing/crossbow/_build/results?buildId=45749&view=logs&j=0da5d1d9-276d-5173-c4c4-9d4d4ed14fdb&t=d9b15392-e4ce-5e4c-0c8c-b69645229181

I suspect those are not caused by this PR - do we have these failures on default branch as well?

**Review by icexelloss — APPROVED** [review]
LGTM

**Comment by icexelloss** [comment]
@westonpace @ildipo I approved. Is there anything you would like to check/change? If not I will merge this end of day today.

**Comment by ildipo** [comment]
nothing else to add to this PR

**Comment by ildipo** [comment]
just don't close the associated issue - I think github will close it automatically

**Comment by assignUser** [comment]
Please create separate issues and track them in an umbrella issue to avoid having to re-open issues to avoid confusion now and during changelog creation. 

I have created #34630 for this PR @ildipo  please comment "take" on it to have it assigned (automatic assignment failed as you have not interacted with the issue - a GH spam protection)

**Comment by kou** [comment]
@icexelloss Could you use https://github.com/apache/arrow/blob/main/dev/merge_arrow_pr.py instead of GitHub's merge button as much as possible? The script does some more preparations.

**Comment by icexelloss** [comment]
Oh sure sorry didn’t realize that. Will certainly do next time.

**Pull request merged**
